lzy/long-running/src/main/java/ai/lzy/longrunning/LocalOperationService.java

```diff
@@ -15,67 +24,258 @@ public class LocalOperationService extends LongRunningServiceGrpc.LongRunningSer
     private static final Logger LOG = LogManager.getLogger(LocalOperationService.class);
 
     private final String name;
+
     private final Map<String, Operation> operations = new ConcurrentHashMap<>();
     private final Map<String, Operation> idempotencyKey2Operation = new ConcurrentHashMap<>();
 
     public LocalOperationService(String name) {
         this.name = name;
     }
 
-    public synchronized boolean registerOperation(Operation operation) {
-        LOG.info("[{}] Register operation {}", name, operation.toShortString());
+    /**
+     * Atomic upsert operation. Returns immutable copy added or already presented operation,
+     * which should be considered as operation's state snapshot at some instant of time.
```

Comment:
snapshot - is already about fixed time point, so you may drop `at some instant of time`